fix tag ordering

francoisverbeek opened the pull request:
fix the tag ordering issue when using FIX44's spec. 
This contains a few formatting/typos fixes in addition to the bug fix changes, let me know if you want me to split these changes out. The test changes pass against the FIX44.xml reference. I'll look at running the tests automatically against multiple specs.

francoisverbeek committed:
fix tag ordering

raoyitao commented on tests/test_ref.py:
```diff
@@ -257,46 +257,17 @@ def test_nested_rgroup(self, spec):
+        serialised = '35=AE;555=1;687=AA;683=2;688=1;689=1;' \
+                     '688=2;689=2;17807=11;10=000;'.replace(';', chr(1)).encode('UTF-8')
         assert serialised == codec.serialise(msg)
 
```
Why remove test function "test_empty_rgroups(self, spec)" ?

francoisverbeek replied:
no reason that I can remember, restored!

raoyitao replied:
A test function **test_from_dict_with_none()** is not found, also 4 lines of function **test_update_with_none()** is not found, I guess you made the change on an older version, maybe you should rebase the code based on the latest version.

Commit:
restore unexpected deleted test case

restore unexpected deleted test case

francoisverbeek commented on tests/test_ref.py:
```diff
@@ -594,7 +594,6 @@ def test_from_buffer(self):
         assert b'D' == msg[35]
         assert b'3' == msg[34]
         assert b'154' == msg[10]
-        assert msg.get_raw_message() == buff
```
checking also why this assertion was removed

francoisverbeek replied:
restored too

Commit:
restore erroneously deleted assertion

restore erroneously deleted assertion

raoyitao commented on pyfixmsg/reference.py:
```diff
@@ -292,20 +312,24 @@ def add_group(self, count_tag, composition, insert_at=None):
+        sorting_key = {35: -1, 10: int(10e9)}
+        for index, item in enumerate(spec.header_tags):
+            sorting_key[item.tag] = index
+    start_index = index + 1
```
Thanks for your PR, I have some comment for it. It seems good for FIX4.2 or FIX4.4, but Quick FIX 5.0 spec has no headers defined (they were defined in another transportation spec, we might support it later). So, I don't think it works for FIX5.0 or later version.
For example, my message is '8=FIX.4.2;9=97;35=B;215=1;216=1;146=2;55=EURUSD;55=EURGBP;10=000;', after parsing it and serializing the message, I found tag 8 and 9 are not in correct order.

Something could be considered:
1). In FixSpec::__init__(), self.header_tags = [...] **or [self.tags.by_tag(tag) for tag in HEADER_TAGS]**
HEADER_TAGS is pre-defined in this file, but it's for FIX4.2 so I'm afraid it needs a little change for FIX5.0, is there any idea to make is available for all versions?
2). HEADER_TAGS of FIX4.4/5.0 also has group, just enumerating spec.header_tags to get tags' index is not enough to handle repeating groups.

francoisverbeek replied:
Good point. So yes, updating ``HEADER_TAGS`` to fit FIXT1.0 solves your example, but doesn't support the Hops repeating group. I think that adding to the doc saying that particular group is not yet supported until such time that someone actively needs it would be fair; what do you think?

raoyitao replied:
Yes, no need to support repeating group in header for now. But this code should work for both FIX4.x and FIX5.x so I suggest the following code, then we could remove if condition in **test_serialisation_header()**

 for index, item in enumerate(header_tags):
            if isinstance(item, FixTag):
                sorting_key[item.tag] = index
            else:
                sorting_key[item] = index

francoisverbeek replied:
You’re right, i’ll remove the if statement. It made me realise that header ordering without an explicit spec is broken, will fix that now.

francoisverbeek replied:
fixed in latest commit

raoyitao commented on .gitignore:
```diff
@@ -9,3 +9,5 @@ fixspecs/FIX40.xml
 
 .idea/pyfixmsg.iml
+
+venv/*
```
Well, why add a venv dir here, is it in your own develop environment ? maybe the fixspecs/FIX40.xml could be removed.

francoisverbeek replied:
removed

franso6 commented:
Good point! Thank you! I’ll add a test case for that, and think about the header tags. 

François

> On 18 Apr 2019, at 19:45, raoyitao <notifications@github.com> wrote:
> 
> @raoyitao commented on this pull request.
> 
> In pyfixmsg/reference.py:
> 
> >      """
>      Retrieve the sorting key for an object.
>      The sorting key is used to serialise tags in the order they appear in the spec.
>      It is unclear whether that's required for the root of the message (aside from header and tail)
>      but it is essential in repeating groups. This takes the safe approach of enforcing it at all
>      levels.
>      """
> -    sorting_key = sorting_key or {35: -1, 10: int(10e9)}
> +    if sorting_key is None:
> +        sorting_key = {35: -1, 10: int(10e9)}
> +        for index, item in enumerate(spec.header_tags):
> +            sorting_key[item.tag] = index
> +    start_index = index + 1
> Thanks for your PR, I have some comment for it. It seems good for FIX4.2 or FIX4.4, but Quick FIX 5.0 spec has no headers defined (they were defined in another transportation spec, we might support it later). So, I don't think it works for FIX5.0 or later version.
> For example, my message is '8=FIX.4.2;9=97;35=B;215=1;216=1;146=2;55=EURUSD;55=EURGBP;10=000;', after parsing it and serializing the message, I found tag 8 and 9 are not in correct order.
> 
> Something could be considered:
> 1). In FixSpec::init(), self.header_tags = [...] or [self.tags.by_tag(tag) for tag in HEADER_TAGS]
> HEADER_TAGS is pre-defined in this file, but it's for FIX4.2 so I'm afraid it needs a little change for FIX5.0, is there any idea to make is available for all versions?
> 2). HEADER_TAGS of FIX4.4/5.0 also has group, just enumerating spec.header_tags to get tags' index is not enough to handle repeating groups.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

franso6 commented:
Yes that’s my virtual environment, will remove. 

François

> On 18 Apr 2019, at 19:50, raoyitao <notifications@github.com> wrote:
> 
> @raoyitao commented on this pull request.
> 
> In .gitignore:
> 
> > @@ -9,3 +9,5 @@ fixspecs/FIX40.xml
>  .idea/.name
>  
>  .idea/pyfixmsg.iml
> +
> +venv/*
> Well, why add a venv dir here, is it in your own develop environment ? maybe the fixspecs/FIX40.xml could be removed.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
fix tag ordering for FIX5.0

and add documentation on not supporting the Hops repeating group inside the header

raoyitao commented on tests/test_ref.py:
```diff
@@ -650,3 +656,12 @@ def test_change_spec(self, spec):
         assert list(after.find_all(270)) == [[268, 0, 270], [268, 1, 270]]
+
+    def test_serialisation_header(self, spec):
+        if 'FIX5' in spec.version:
```
If condition could be remove because each version of fix spec should support tag order of header.

raoyitao commented on tests/test_ref.py:
```diff
@@ -650,3 +656,12 @@ def test_change_spec(self, spec):
         with pytest.raises(KeyError):
             after[270]
         assert list(after.find_all(270)) == [[268, 0, 270], [268, 1, 270]]
+
```
I guess you made the change based on an older version of pyfixmsg, maybe you could rebase to the latest. I found the test function **test_from_dict_with_none()** is not found, also 4 lines are missing in function **test_update_with_none()**.

francoisverbeek replied:
I’ll check, I didn’t notice them missing in the review on the pull request..

francoisverbeek replied:
I don't see changes in the public repo for these -- could these be changes that are available in your private repo?

raoyitao replied:
As I checked again you don't need to rebase, because I got a new edition of code, sorry for that,

Commit:
fix types in header tags

sorting key uses all ints -- the header tags were Tags

raoyitao commented on pyfixmsg/reference.py:
```diff
@@ -175,7 +175,7 @@ def __init__(self, xml_file, eager=False):
         self.msg_types.update({m.msgtype: m for m in
                                (MessageType(e, self) for e in self.tree.findall('messages/message'))})
-        self.header_tags = [self.tags.by_name(t.get('name')) for t in self.tree.findall('header/field')]
+        self.header_tags = [self.tags.by_name(t.get('name')).tag for t in self.tree.findall('header/field')]
```
It works actually. But I still have a concern. because some of our internal project also includes pyfixmsg as a module, so I'm afraid that change the data type of FixSpec.header_tags from [FixTag] to [int] might lead to some incompatibility issue. Also, the constant HEADER_TAGS is refered by others. So I suggest that in function **_extract_sorting_key()**:

header_tags = spec.header_tags or [self.tags.by_name(tag) for tag in HEADER_TAGS]
for index, item in enumerate(header_tags):
    sorting_key[item.tag] = index

I think the PR is almost done.  ^_^

francoisverbeek replied:
you're right, fixed.

Commit:
restore spec.header_tags type

retain the original type for self.header_tags in Spec, and change type only when used in sorting_key
